x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/SpanFlyout/TruncateHeightSection.tsx

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { EuiLink } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import React, { Fragment, useEffect, useRef, useState } from 'react';
+import styled from 'styled-components';
+import { idx } from 'x-pack/plugins/apm/common/idx';
+import { px, units } from '../../../../../../../style/variables';
+// @ts-ignore
+import { Ellipsis } from '../../../../../../shared/Icons';
+
+const ToggleButtonContainer = styled.div`
+  margin-top: ${px(units.half)};
+  user-select: none;
+`;
+
+const ContentContainer = styled.div`
+  overflow: hidden;
+`;
+
+interface Props {
+  previewHeight: string;
+}
+
+export const TruncateHeightSection: React.SFC<Props> = props => {
+  const [isTruncated, setIsTruncated] = useState(true);
+  const [showToggle, setShowToggle] = useState(false);
+  const contentContainerEl = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    const scrollHeight =
+      idx(contentContainerEl, _ => _.current.scrollHeight) || 0;
+    const clientHeight =
+      idx(contentContainerEl, _ => _.current.clientHeight) || 0;
+    if (scrollHeight > clientHeight) {
+      setShowToggle(true);
+    } else {
+      setIsTruncated(false);
+    }
```

Comment:
There seems to be a logical error here: if initially `scrollHeight` is greater than `clientHeight`, the toggle will be displayed. But if props are updated, so either children is decreased in height, or `previewHeight` is increased,  `showToggle` should be set to `false`. But it isn't. Since state is persisted between renders, showToggle will always be displayed regardless if props change.

Reply:
ok i get what ur saying, and i think i can overcome it with `setShowToggle(scrollHeight > clientHeight);` in the effect hook so that it will be set to false by default if it's triggered.

Reply:
I tried this a number of ways, but the only way i can get a reliable value for `showToggle` is to force the component to render collapsed at least once between updates (adds one more bit of state) to check if there is clipping or not. It brings back some of the complicated issues as before. (requires 2 renders per change in props).

Another way i can think of is to have an invisible container div that is always set to the `previewHeight` regardless of state, and to check clipping on that element on updates.

I'll post the 2-render solution i have first and tell me what u think, if not we can explore the other solution.